core/cloudflow-operator/src/main/scala/cloudflow/operator/CloudflowApplication.scala

```diff
@@ -154,30 +195,47 @@ object CloudflowApplication {
         streamletStatuses
           .find(_.streamletName == streamletName)
           .map(_.updatePod(pod))
-          .getOrElse(CloudflowApplication.StreamletStatus(streamletName, pod))
-      copy(streamletStatuses = streamletStatuses.filterNot(_.streamletName == streamletStatus.streamletName) :+ streamletStatus)
+          .toList
+      val streamletStatusesUpdated = streamletStatuses.filterNot(_.streamletName == streamletName) ++ streamletStatus
+      copy(
+        streamletStatuses = streamletStatusesUpdated,
+        appStatus = Status.calcAppStatus(streamletStatusesUpdated)
+      )
     }
 
     def deletePod(streamletName: String, pod: Pod) = {
       val streamletStatus =
         streamletStatuses
           .find(_.streamletName == streamletName)
           .map(_.deletePod(pod))
-          .getOrElse(CloudflowApplication.StreamletStatus(streamletName, pod))
-      copy(streamletStatuses = streamletStatuses.filterNot(_.streamletName == streamletStatus.streamletName) :+ streamletStatus)
+          .toList
+      val streamletStatusesUpdated = streamletStatuses.filterNot(_.streamletName == streamletName) ++ streamletStatus
+      copy(
+        streamletStatuses = streamletStatusesUpdated,
+        appStatus = Status.calcAppStatus(streamletStatusesUpdated)
+      )
     }
 
     def toAction(app: CloudflowApplication.CR): Action[ObjectResource] =
       Action.updateStatus(app.withStatus(this), editor)
   }
 
   object StreamletStatus {
-    def apply(streamletName: String, pod: Pod): StreamletStatus =
-      StreamletStatus(streamletName, Vector(PodStatus(pod)))
-    def apply(streamletName: String): StreamletStatus = StreamletStatus(streamletName)
+    def apply(streamletName: String, expectedPodCount: Int): StreamletStatus =
+      StreamletStatus(streamletName, Some(expectedPodCount), Vector())
+    def apply(streamletName: String, pod: Pod, expectedPodCount: Int): StreamletStatus =
+      StreamletStatus(streamletName, Some(expectedPodCount), Vector(PodStatus(pod)))
+    def apply(streamletName: String, expectedPodCount: Int, podStatuses: Vector[PodStatus]): StreamletStatus =
+      StreamletStatus(streamletName, Some(expectedPodCount), podStatuses)
   }
 
-  case class StreamletStatus(streamletName: String, podStatuses: Vector[PodStatus] = Vector()) {
+  case class StreamletStatus(
+      streamletName: String,
+      // Added as Option for backwards compatibility
+      expectedPodCount: Option[Int],
+      podStatuses: Vector[PodStatus]
+  ) {
+    def hasExpectedPods(nrOfPodsDetected: Int) = expectedPodCount.getOrElse(0) == nrOfPodsDetected
```

Comment:
How can `expectedPodCount` be `None` - I would think that the expected count is always known, no? (while `actual` might be unknown.)

Reply:
@maasg The issue is that existing apps don't have this field. If I make it mandatory the operator will crash. (cause it can't read the status without the field). So this is a best of worst cases, instead of the operator crashing on existing apps, it will badly report the status of those apps. We'll document this in the release and tell users to redeploy. If they don't read those notes their stuff will at least not crash.

Reply:
@maasg the same applies for nrOfContainers and nrOfContainersReady in PodStatus.. Welcome to the world of backwards compat by optional fields ;-)